client/hwlib/src/collectors/cpuinfo.rs

```diff
@@ -0,0 +1,183 @@
+pub fn parse_cpuinfo(
+    cpuinfo_filepath: Option<&'static str>,
+) -> Result<CpuInfo, Box<dyn std::error::Error>> {
+    let path = cpuinfo_filepath.unwrap_or(PROC_CPUINFO_FILEPATH);
+    let file = File::open(path)?;
+    let reader = io::BufReader::new(file);
+
+    let mut attributes: HashMap<String, String> = HashMap::new();
+    let mut count = 0;
+
+    for line in reader.lines() {
+        let line = line?;
+        if line.trim().is_empty() {
+            continue;
+        }
+
+        let parts: Vec<&str> = line.split(':').collect();
+        if parts.len() == 2 {
+            let key = parts[0].trim().to_string();
+            let value = parts[1].trim().to_string();
+
+            if key == "processor" {
+                count += 1;
+            }
+
+            attributes.insert(key, value);
+        }
+    }
+
+    let machine = std::env::consts::ARCH.to_string();
+    let speed = parse_speed(attributes.get("cpu MHz")).unwrap_or(0);
+
+    let model = attributes
+        .get("Model")
+        .or_else(|| attributes.get("model name"))
+        .unwrap_or(&machine)
+        .to_string();
+
+    let cpu_info = CpuInfo {
+        platform: machine.clone(),
+        count,
+        cpu_type: attributes.get("vendor_id").unwrap_or(&machine).to_string(),
+        model,
+        model_number: attributes
+            .get("cpu family")
+            .unwrap_or(&"".to_string())
+            .to_string(),
+        model_version: attributes
+            .get("model")
+            .unwrap_or(&"".to_string())
+            .to_string(),
+        model_revision: attributes
+            .get("stepping")
+            .unwrap_or(&"".to_string())
+            .to_string(),
+        cache: parse_cache_size(attributes.get("cache size"))?,
+        bogomips: parse_bogomips(attributes.get("bogomips"))?,
+        speed,
+        other: attributes
+            .get("flags")
+            .unwrap_or(&"".to_string())
+            .to_string(),
+    };
+
+    Ok(cpu_info)
+}
+
+/// Parse CPU frequency in MHz as it's done in checkbox
+/// https://github.com/canonical/checkbox/blob/a8d5e9/providers/resource/bin/cpuinfo_resource.py#L56-L63
+pub(super) fn read_max_cpu_frequency(
+    max_cpu_frequency_filepath: Option<&'static str>,
+) -> Result<u64, Box<dyn std::error::Error>> {
+    let path = max_cpu_frequency_filepath.unwrap_or(CPU_MAX_FREQ_FILEPATH);
+    let file = File::open(path)?;
+    let mut reader = io::BufReader::new(file);
+    let mut buffer = String::new();
+    reader.read_line(&mut buffer)?;
+    let k_hz: u64 = buffer.trim().parse()?;
+    Ok(k_hz / 1000) // Convert kHz to MHz
+}
+
+fn parse_cache_size(cache_size: Option<&String>) -> Result<i64, Box<dyn std::error::Error>> {
+    if let Some(cache) = cache_size {
+        let cache_str = cache.replace(" KB", "");
+        return i64::from_str(&cache_str).map_err(|e| e.into());
+    }
+    Ok(-1)
```

Comment:
For each of these parsing functions I'd suggest not accepting optional inputs: instead only parse the value if it is present and set the default above in `parse_cpuinfo`. Unless these sentinel values are required for the data format due to existing API restrictions, I'd also recommend having the values as `Options` to indicate where values are missing.

As a general rule, in Rust it is encouraged to avoid "magic values" wherever possible.

Reply:
updated in cddb04d